File: swim-java/swim-core/swim.util/src/main/java/swim/util/Severity.java
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.util;

import java.util.Objects;
import swim.annotations.Nullable;
import swim.annotations.Public;
import swim.annotations.Since;
import swim.decl.Marshal;
import swim.decl.Unmarshal;

/**
 * Discrete severity scale used for logging and diagnostic classification.
 */
@Public
@Since("5.0")
public final class Severity implements Comparable<Severity>, ToMarkup, ToSource {

  final int value;
  final String label;

  private Severity(int value, String label) {
    this.value = value;
    this.label = label;
  }

  /**
   * Returns the ordinal severity value, with higher integral values
   * signifying increased severity.
   *
   * @return an integer between {@code 0} and {@code 8}, inclusive.
   *         One of {@code ALL_VALUE}, {@code TRACE_VALUE}, {@code DEBUG_VALUE},
   *         {@code INFO_VALUE}, {@code NOTICE_VALUE}, {@code WARNING_VALUE},
   *         {@code ERROR_VALUE}, {@code FATAL_VALUE}, or {@code OFF_VALUE}
   */
  public int value() {
    return this.value;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #ALL_VALUE}.
   */
  public boolean isAll() {
    return this.value == ALL_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #TRACE_VALUE}.
   */
  public boolean isTrace() {
    return this.value == TRACE_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #DEBUG_VALUE}.
   */
  public boolean isDebug() {
    return this.value == DEBUG_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #INFO_VALUE}.
   */
  public boolean isInfo() {
    return this.value == INFO_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #NOTICE_VALUE}.
   */
  public boolean isNotice() {
    return this.value == NOTICE_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #WARNING_VALUE}.
   */
  public boolean isWarning() {
    return this.value == WARNING_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #ERROR_VALUE}.
   */
  public boolean isError() {
    return this.value == ERROR_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #FATAL_VALUE}.
   */
  public boolean isFatal() {
    return this.value == FATAL_VALUE;
  }

  /**
   * Returns {@code true} if the {@linkplain #value() severity value}
   * is equal to {@link #OFF_VALUE}.
   */
  public boolean isOff() {
    return this.value == OFF_VALUE;
  }

  /**
   * Returns a descriptive label for this severity.
   */
  public String label() {
    return this.label;
  }

  /**
   * Returns a new {@code Severity} with the same severity value,
   * but with a new descriptive {@code label}.
   */
  public Severity withLabel(String label) {
    if (this.label.equals(label)) {
      return this;
    }
    return Severity.of(this.value, label);
  }

  public boolean filter(Severity that) {
    return this.value <= that.value;
  }

  public Severity min(Severity that) {
    if (this.value <= that.value) {
      return this;
    } else {
      return that;
    }
  }

  public Severity max(Severity that) {
    if (this.value >= that.value) {
      return this;
    } else {
      return that;
    }
  }

  @Override
  public int compareTo(Severity that) {
    if (this == that) {
      return 0;
    } else if (this.value < that.value) {
      return -1;
    } else if (this.value > that.value) {
      return 1;
    }
    return this.label.compareTo(that.label);
  }

  @Override
  public boolean equals(@Nullable Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof Severity that) {
      return this.value == that.value && this.label.equals(that.label);
    }
    return false;
  }

  private static final int HASH_SEED = Murmur3.seed(Severity.class);

  @Override
  public int hashCode() {
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(HASH_SEED,
        this.value), this.label.hashCode()));
  }

  public void stylize(Notation notation) {
    switch (this.value) {
      case FATAL_VALUE:
        notation.boldMagenta();
        break;
      case ERROR_VALUE:
        notation.boldRed();
        break;
      case WARNING_VALUE:
        notation.boldYellow();
        break;
      case NOTICE_VALUE:
        notation.boldGreen();
        break;
      case INFO_VALUE:
        notation.boldBlue();
        break;
      case DEBUG_VALUE:
        notation.boldCyan();
        break;
      case TRACE_VALUE:
      default:
        notation.boldGray();
    }
  }

  @Override
  public void writeSource(Appendable output) {
    final Notation notation = Notation.from(output);
    notation.append("Severity").append('.');
    final boolean withLabel;
    switch (this.value) {
      case ALL_VALUE:
        notation.append("ALL");
        withLabel = !ALL_LABEL.equals(this.label);
        break;
      case TRACE_VALUE:
        notation.append("TRACE");
        withLabel = !TRACE_LABEL.equals(this.label);
        break;
      case DEBUG_VALUE:
        notation.append("DEBUG");
        withLabel = !DEBUG_LABEL.equals(this.label);
        break;
      case INFO_VALUE:
        notation.append("INFO");
        withLabel = !INFO_LABEL.equals(this.label);
        break;
      case NOTICE_VALUE:
        notation.append("NOTICE");
        withLabel = !NOTICE_LABEL.equals(this.label);
        break;
      case WARNING_VALUE:
        notation.append("WARNING");
        withLabel = !WARNING_LABEL.equals(this.label);
        break;
      case ERROR_VALUE:
        notation.append("ERROR");
        withLabel = !ERROR_LABEL.equals(this.label);
        break;
      case FATAL_VALUE:
        notation.append("FATAL");
        withLabel = !FATAL_LABEL.equals(this.label);
        break;
      case OFF_VALUE:
        notation.append("OFF");
        withLabel = !OFF_LABEL.equals(this.label);
        break;
      default:
        throw new AssertionError(Integer.toString(this.value));
    }
    if (withLabel) {
      notation.beginInvoke("withLabel").appendArgument(this.label).endInvoke();
    }
  }

  @Override
  public void writeMarkup(Appendable output) {
    final Notation notation = Notation.from(output);
    this.stylize(notation);
    notation.appendMarkup(this.label);
    notation.reset();
  }

  @Override
  @Marshal
  public String toString() {
    return this.label;
  }

  public static final int ALL_VALUE = 0;
  public static final int TRACE_VALUE = 1;
  public static final int DEBUG_VALUE = 2;
  public static final int INFO_VALUE = 3;
  public static final int NOTICE_VALUE = 4;
  public static final int WARNING_VALUE = 5;
  public static final int ERROR_VALUE = 6;
  public static final int FATAL_VALUE = 7;
  public static final int OFF_VALUE = 8;

  static final String ALL_LABEL = "all";
  static final String TRACE_LABEL = "trace";
  static final String DEBUG_LABEL = "debug";
  static final String INFO_LABEL = "info";
  static final String NOTICE_LABEL = "notice";
  static final String WARNING_LABEL = "warning";
  static final String ERROR_LABEL = "error";
  static final String FATAL_LABEL = "fatal";
  static final String OFF_LABEL = "off";

  public static final Severity ALL = new Severity(ALL_VALUE, ALL_LABEL);
  public static final Severity TRACE = new Severity(TRACE_VALUE, TRACE_LABEL);
  public static final Severity DEBUG = new Severity(DEBUG_VALUE, DEBUG_LABEL);
  public static final Severity INFO = new Severity(INFO_VALUE, INFO_LABEL);
  public static final Severity NOTICE = new Severity(NOTICE_VALUE, NOTICE_LABEL);
  public static final Severity WARNING = new Severity(WARNING_VALUE, WARNING_LABEL);
  public static final Severity ERROR = new Severity(ERROR_VALUE, ERROR_LABEL);
  public static final Severity FATAL = new Severity(FATAL_VALUE, FATAL_LABEL);
  public static final Severity OFF = new Severity(OFF_VALUE, OFF_LABEL);

  /**
   * Returns a {@code Severity} with the given {@code value}
   * and descriptive {@code label}.
   *
   * @throws IllegalArgumentException if {@code value} is not
   *         a valid severity value.
   */
  public static Severity of(int value, @Nullable String label) {
    switch (value) {
      case ALL_VALUE:
        if (label == null || ALL_LABEL.equals(label)) {
          return ALL;
        }
        break;
      case TRACE_VALUE:
        if (label == null || TRACE_LABEL.equals(label)) {
          return TRACE;
        }
        break;
      case DEBUG_VALUE:
        if (label == null || DEBUG_LABEL.equals(label)) {
          return DEBUG;
        }
        break;
      case INFO_VALUE:
        if (label == null || INFO_LABEL.equals(label)) {
          return INFO;
        }
        break;
      case NOTICE_VALUE:
        if (label == null || NOTICE_LABEL.equals(label)) {
          return NOTICE;
        }
        break;
      case WARNING_VALUE:
        if (label == null || WARNING_LABEL.equals(label)) {
          return WARNING;
        }
        break;
      case ERROR_VALUE:
        if (label == null || ERROR_LABEL.equals(label)) {
          return ERROR;
        }
        break;
      case FATAL_VALUE:
        if (label == null || FATAL_LABEL.equals(label)) {
          return FATAL;
        }
        break;
      case OFF_VALUE:
        if (label == null || OFF_LABEL.equals(label)) {
          return OFF;
        }
        break;
      default:
        throw new IllegalArgumentException("invalid severity value: " + value);
    }
    return new Severity(value, label);
  }

  /**
   * Returns a {@code Severity} with the given {@code value}.
   *
   * @throws IllegalArgumentException if {@code value} is not
   *         a valid severity value.
   */
  public static Severity of(int value) {
    switch (value) {
      case ALL_VALUE:
        return ALL;
      case TRACE_VALUE:
        return TRACE;
      case DEBUG_VALUE:
        return DEBUG;
      case INFO_VALUE:
        return INFO;
      case NOTICE_VALUE:
        return NOTICE;
      case WARNING_VALUE:
        return WARNING;
      case ERROR_VALUE:
        return ERROR;
      case FATAL_VALUE:
        return FATAL;
      case OFF_VALUE:
        return OFF;
      default:
        throw new IllegalArgumentException("invalid severity value: " + value);
    }
  }

  /**
   * Returns the {@code Severity} corresponding to the the given {@code label}.
   *
   * @throws IllegalArgumentException if {@code label} is not
   *         a standard severity label.
   */
  @Unmarshal
  public static Severity parse(String label) {
    Objects.requireNonNull(label);
    if (ALL_LABEL.equalsIgnoreCase(label)) {
      return ALL;
    } else if (TRACE_LABEL.equalsIgnoreCase(label)) {
      return TRACE;
    } else if (DEBUG_LABEL.equalsIgnoreCase(label)) {
      return DEBUG;
    } else if (INFO_LABEL.equalsIgnoreCase(label)) {
      return INFO;
    } else if (NOTICE_LABEL.equalsIgnoreCase(label)) {
      return NOTICE;
    } else if (WARNING_LABEL.equalsIgnoreCase(label)) {
      return WARNING;
    } else if (ERROR_LABEL.equalsIgnoreCase(label)) {
      return ERROR;
    } else if (FATAL_LABEL.equalsIgnoreCase(label)) {
      return FATAL;
    } else if (OFF_LABEL.equalsIgnoreCase(label)) {
      return OFF;
    }
    throw new IllegalArgumentException("unknown severity label: " + label);
  }

}

File: swim-js/swim-core/swim-util/main/Severity.ts
// Copyright 2015-2024 Nstream, inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Murmur3} from "./Murmur3";
import type {HashCode} from "./HashCode";
import type {Compare} from "./Compare";
import {Numbers} from "./Numbers";
import {Strings} from "./Strings";
import {Constructors} from "./Constructors";

/**
 * Level of importance. Used for log levels and diagnostic classifications.
 * @public
 */
export class Severity implements HashCode, Compare {
  /** @internal */
  constructor(level: number, label: string) {
    this.level = level;
    this.label = label;
  }

  /**
   * The integer level of importance of this `Severity`, with higher  levels
   * signifying greater importance. An integer between `0` and `7`, inclusive.
   * One of `TraceLevel`, `DebugLevel`, `InfoLevel`, `NoteLevel`,
   * `WarningLevel`, `ErrorLevel`, `AlertLevel`, `FatalLevel`.
   */
  readonly level: number;

  /**
   * A descriptive label for this `Severity`.
   */
  readonly label: string;

  /**
   * Returns `true` if this `Severity` has `TraceLevel` of importance.
   */
  isTrace(): boolean {
    return this.level === Severity.TraceLevel;
  }

  /**
   * Returns `true` if this `Severity` has `DebugLevel` of importance.
   */
  isDebug(): boolean {
    return this.level === Severity.DebugLevel;
  }

  /**
   * Returns `true` if this `Severity` has `InfoLevel` of importance.
   */
  isInfo(): boolean {
    return this.level === Severity.InfoLevel;
  }

  /**
   * Returns `true` if this `Severity` has `NoteLevel` of importance.
   */
  isNote(): boolean {
    return this.level === Severity.NoteLevel;
  }

  /**
   * Returns `true` if this `Severity` has `WarningLevel` of importance.
   */
  isWarning(): boolean {
    return this.level === Severity.WarningLevel;
  }

  /**
   * Returns `true` if this `Severity` has `ErrorLevel` of importance.
   */
  isError(): boolean {
    return this.level === Severity.ErrorLevel;
  }

  /**
   * Returns `true` if this `Severity` has `AlertLevel` of importance.
   */
  isAlert(): boolean {
    return this.level === Severity.AlertLevel;
  }

  /**
   * Returns `true` if this `Severity` has `FatalLevel` of importance.
   */
  isFatal(): boolean {
    return this.level === Severity.FatalLevel;
  }

  compareTo(that: unknown): number {
    if (this === that) {
      return 0;
    } else if (that instanceof Severity) {
      if (this.level < that.level) {
        return -1;
      } else if (this.level > that.level) {
        return 1;
      }
      return this.label.localeCompare(that.label);
    }
    return NaN;
  }

  equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof Severity) {
      return this.level === that.level && this.label === that.label;
    }
    return false;
  }

  hashCode(): number {
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(Constructors.hash(Severity),
        Numbers.hash(this.level)), Strings.hash(this.label)));
  }

  toString(): string {
    return this.label;
  }

  static readonly TraceLevel: number = 0;
  static readonly DebugLevel: number = 1;
  static readonly InfoLevel: number = 2;
  static readonly NoteLevel: number = 3;
  static readonly WarningLevel: number = 4;
  static readonly ErrorLevel: number = 5;
  static readonly AlertLevel: number = 6;
  static readonly FatalLevel: number = 7;

  static readonly TraceLabel: string = "trace";
  static readonly DebugLabel: string = "debug";
  static readonly InfoLabel: string = "info";
  static readonly NoteLabel: string = "note";
  static readonly WarningLabel: string = "warning";
  static readonly ErrorLabel: string = "error";
  static readonly AlertLabel: string = "alert";
  static readonly FatalLabel: string = "fatal";

  /** @internal */
  static Trace: Severity | null = null;
  /** @internal */
  static Debug: Severity | null = null;
  /** @internal */
  static Info: Severity | null = null;
  /** @internal */
  static Note: Severity | null = null;
  /** @internal */
  static Warning: Severity | null = null;
  /** @internal */
  static Error: Severity | null = null;
  /** @internal */
  static Alert: Severity | null = null;
  /** @internal */
  static Fatal: Severity | null = null;

  /**
   * Returns a `Severity` with the given importance `level`, and optional
   * descriptive `label`.
   *
   * @throws `Error` if `level` is not a valid level of importance.
   */
  static create(level: number, label?: string): Severity {
    switch (level) {
      case this.TraceLevel:
        return this.trace(label);
      case this.DebugLevel:
        return this.debug(label);
      case this.InfoLevel:
        return this.info(label);
      case this.NoteLevel:
        return this.note(label);
      case this.WarningLevel:
        return this.warning(label);
      case this.ErrorLevel:
        return this.error(label);
      case this.AlertLevel:
        return this.alert(label);
      case this.FatalLevel:
        return this.fatal(label);
      default:
        throw new Error("" + level);
    }
  }

  /**
   * Returns a `Severity` with `TraceLevel` of importance, and an optional
   * descriptive `label`.
   */
  static trace(label?: string): Severity {
    if (label === void 0 || label === this.TraceLabel) {
      if (this.Trace === null) {
        this.Trace = new Severity(this.TraceLevel, this.TraceLabel);
      }
      return this.Trace;
    }
    return new Severity(this.TraceLevel, label);
  }

  /**
   * Returns a `Severity` with `DebugLevel` of importance, and an optional
   * descriptive `label`.
   */
  static debug(label?: string): Severity {
    if (label === void 0 || label === this.DebugLabel) {
      if (this.Debug === null) {
        this.Debug = new Severity(this.DebugLevel, this.DebugLabel);
      }
      return this.Debug;
    }
    return new Severity(this.DebugLevel, label);
  }

  /**
   * Returns a `Severity` with `InfoLevel` of importance, and an optional
   * descriptive `label`.
   */
  static info(label?: string): Severity {
    if (label === void 0 || label === this.InfoLabel) {
      if (this.Info === null) {
        this.Info = new Severity(this.InfoLevel, this.InfoLabel);
      }
      return this.Info;
    }
    return new Severity(this.InfoLevel, label);
  }

  /**
   * Returns a `Severity` with `NoteLevel` of importance, and an optional
   * descriptive `label`.
   */
  static note(label?: string): Severity {
    if (label === void 0 || label === this.NoteLabel) {
      if (this.Note === null) {
        this.Note = new Severity(this.NoteLevel, this.NoteLabel);
      }
      return this.Note;
    }
    return new Severity(this.NoteLevel, label);
  }

  /**
   * Returns a `Severity` with `WarningLevel` of importance, and an optional
   * descriptive `label`.
   */
  static warning(label?: string): Severity {
    if (label === void 0 || label === this.WarningLabel) {
      if (this.Warning === null) {
        this.Warning = new Severity(this.WarningLevel, this.WarningLabel);
      }
      return this.Warning;
    }
    return new Severity(this.WarningLevel, label);
  }

  /**
   * Returns a `Severity` with `ErrorLevel` of importance, and an optional
   * descriptive `label`.
   */
  static error(label?: string): Severity {
    if (label === void 0 || label === this.ErrorLabel) {
      if (this.Error === null) {
        this.Error = new Severity(this.ErrorLevel, this.ErrorLabel);
      }
      return this.Error;
    }
    return new Severity(this.ErrorLevel, label);
  }

  /**
   * Returns a `Severity` with `AlertLevel` of importance, and an optional
   * descriptive `label`.
   */
  static alert(label?: string): Severity {
    if (label === void 0 || label === this.AlertLabel) {
      if (this.Alert === null) {
        this.Alert = new Severity(this.AlertLevel, this.AlertLabel);
      }
      return this.Alert;
    }
    return new Severity(this.AlertLevel, label);
  }

  /**
   * Returns a `Severity` with `FatalLevel` of importance, and an optional
   * descriptive `label`.
   */
  static fatal(label?: string): Severity {
    if (label === void 0 || label === this.FatalLabel) {
      if (this.Fatal === null) {
        this.Fatal = new Severity(this.FatalLevel, this.FatalLabel);
      }
      return this.Fatal;
    }
    return new Severity(this.FatalLevel, label);
  }
}
